src/rsa/key_generation.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2018 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! RSA key generation
+
+use super::small_primes::SMALL_PRIMES;
+use core::marker::PhantomData;
+use std::vec::Vec;
+
+use crate::{
+    limb::{self, LimbMask},
+    error, rand,
+    rsa::bigint,
+};
+
+/// Performs the Miller-Rabin primality test
+///
+/// Implementation of the algorithm as it
+/// is described in FIPS 186-4 C.3.1
+fn miller_rabin_test(
+    w: &[u64],
+    iterations: u32,
+    rng: &rand::SecureRandom,
+) -> Result<bool, error::Unspecified> {
+    // Step 1.
+    let mut w_m1 = vec![0; w.len()];
+    limb::limbs_sub_limb(&mut w_m1, w, 1);
+    let a = limb::limbs_count_low_zero_bits(&w_m1);
+
+    // Step 2.
+    let mut m = vec![0; w_m1.len()];
+    limb::limbs_copy(&mut m, &w_m1);
+    limb::limbs_rshift(&mut m, a);
+
+    // Step 3 not needed.
+
+    // preparation
+    let boxed_w = bigint::BoxedLimbs::<()>::minimal_width_from_unpadded(&w);
+    let w_modulus = bigint::Modulus::from_boxed_limbs(boxed_w)?.0;
+    let boxed_m = bigint::BoxedLimbs::<()>::minimal_width_from_unpadded(&m);
+    let m_exponent = bigint::PrivateExponent {
+        limbs: boxed_m,
+    };
+    // Step 4.
+    let mut b: Vec<limb::Limb> = vec![0; w.len()];
+    for _ in 0 .. iterations {
+        loop {
+            // Step 4.1
+            unsafe {
+                let b_mut_ptr = (&mut b[..]).as_mut_ptr() as * mut u8;
+                let b_mut_slice = core::slice::from_raw_parts_mut(b_mut_ptr, b.len() * 8);
+                rng.fill(b_mut_slice)?;
```

Comment:
This should probably be factored out into a `impl Elem { fn  random(rng: &SecureRandom) -> Result<Self, error::Unspecified> }` method.